src/ch9329.cpp

```diff
@@ -0,0 +1,401 @@
+{
+  m_data[2] = p_x_offset;
+  m_data[3] = p_y_offset;
+  return *this;
+}
+
+ch9329::mouse_relative& ch9329::mouse_relative::scroll_position(
+  std::int8_t p_scroll_offset)
+{
+  m_data[4] = p_scroll_offset;
+  return *this;
+}
+
+ch9329::mouse_relative& ch9329::mouse_relative::left_button(bool p_pressed)
+{
+  constexpr auto left_button_mask = hal::bit_mask::from<0>();
+  hal::bit_modify(m_data[1]).insert<left_button_mask>(p_pressed);
+  return *this;
+}
+
+ch9329::mouse_relative& ch9329::mouse_relative::middle_button(bool p_pressed)
+{
+  constexpr auto middle_button_mask = hal::bit_mask::from<2>();
+  hal::bit_modify(m_data[1]).insert<middle_button_mask>(p_pressed);
+  return *this;
+}
+
+ch9329::mouse_relative& ch9329::mouse_relative::right_button(bool p_pressed)
+{
+  constexpr auto right_button_mask = hal::bit_mask::from<1>();
+  hal::bit_modify(m_data[1]).insert<right_button_mask>(p_pressed);
+  return *this;
+}
+
+// keyboard media functions
+ch9329::keyboard_media& ch9329::keyboard_media::press_media_key(media_key p_key)
+{
+  auto byte_num = (static_cast<uint8_t>(p_key) >> 3) + 1;
+  auto bit_num = static_cast<uint8_t>(p_key) & 0b111;
+  m_data[byte_num] |= 1 << bit_num;
+  return *this;
+}
+
+ch9329::keyboard_media& ch9329::keyboard_media::release_media_key(
+  media_key p_key)
+{
+  auto byte_num = (static_cast<uint8_t>(p_key) >> 3) + 1;
+  auto bit_num = static_cast<uint8_t>(p_key) & 0b111;
+  m_data[byte_num] &= ~(1 << bit_num);
+  return *this;
+}
+
+ch9329::keyboard_media& ch9329::keyboard_media::release_all_keys()
+{
+  m_data = { 0x01, 0x00, 0x00, 0x00 };
+  return *this;
+}
+
+// keyboard acpi functions
+ch9329::keyboard_acpi& ch9329::keyboard_acpi::press_acpi_key(acpi_key p_key)
+{
+  auto bit_num = static_cast<uint8_t>(p_key);
+  m_data[1] |= 1 << bit_num;
+  return *this;
+}
+
+ch9329::keyboard_acpi& ch9329::keyboard_acpi::release_acpi_key(acpi_key p_key)
+{
+  auto bit_num = static_cast<uint8_t>(p_key);
+  m_data[1] &= ~(1 << bit_num);
+  return *this;
+}
+
+ch9329::keyboard_acpi& ch9329::keyboard_acpi::release_all_keys()
+{
+  m_data = { 0x02, 0x00 };
+  return *this;
+}
+
+// keyboard general functions
+ch9329::keyboard_general& ch9329::keyboard_general::press_control_key(
+  control_key_bit p_key)
+{
+  hal::byte mask = 0x00;
+  switch (p_key) {
+    case control_key_bit::left_control:
+      mask = 0b1;
+      break;
+    case control_key_bit::left_shift:
+      mask = 0b10;
+      break;
+    case control_key_bit::left_alt:
+      mask = 0b100;
+      break;
+    case control_key_bit::left_windows:
+      mask = 0b1000;
+      break;
+    case control_key_bit::right_control:
+      mask = 0b10000;
+      break;
+    case control_key_bit::right_shift:
+      mask = 0b100000;
+      break;
+    case control_key_bit::right_alt:
+      mask = 0b1000000;
+      break;
+    case control_key_bit::right_windows:
+      mask = 0b10000000;
+      break;
+  }
+  m_data[0] = m_data[0] | mask;
```

Comment:
The value of each of these control bits is the bit position, so we can replace this with 

```C++
(1 << hal::value(p_key));
```

And reduce the code size of a switch case block. May be useful to add a comment to the enum class about the values meaning bit positions.